### PR TITLE
[SIP 6] Querycontext improvement

### PR DESCRIPTION
**Pull request by conglei**
This pr is to push my change to the feature branch so it won't bring any trouble to master branch. 

@williaster @xtinec @kristw

### CODE REVIEW

**Comment by conglei** [comment]
Current issue is about the format of metric. 

In the new QueryObject, we want to unify the format of metric and ad-hoc metric. However, it brings:

1) The inconsistent logic of creating queryobject in the backend and frontend;

2) the inconsistent metric format between the new queryobject and form_data. 

The easiest approach is to make the backend (query engine) more tolerant on the difference. But in a term, we need to unify all of them.  (forming queryobject in the backend will be deprecated any way),

**Comment by mistercrunch** [comment]
I know we "squash & merge" in this repo so it matters much less, but still would recommend reading this https://chris.beams.io/posts/git-commit/

File: superset/common/query_context.py [hunk 1]
```diff
@@ -1,27 +1,244 @@
...
 
-    def get_data(self):
-        raise NotImplementedError()
+        self.force = force
+
+        self.custom_cache_timeout = custom_cache_timeout
+
+        self.enforce_numerical_metrics = True
+
+    def get_query_result(self, query_object):
```
Comment (mistercrunch) [inline superset/common/query_context.py]: I think a fair amount of the code here is copied from `viz.py`, reading the PR it's hard to know whether it's been just copied or copied and then altered. Even if the goal is to ultimately deprecate this logic from `viz.py` we should still duplicate as little code as possible in the meantime. Could we do this with a mixin class of some sort?

**Comment by mistercrunch** [comment]
Oh I just read that this is targeting a feature branch...

File: superset/common/query_context.py (hunk 1 above)
Comment (conglei) [reply superset/common/query_context.py]: Yes, I agree that there should be as little duplicated code as possible.  Basically this branch is more for a POC. This branch won't be merged master until this issue is resolved.

**Comment by conglei** [comment]
@williaster  Can we first merge this pr? so it will be easy for me to send a second PR about tablevis to this branch.

**Comment by williaster** [comment]
Yep! merging now.

**Pull request merged**